### ISSUE TITLE
Add WASI HTTP example 

### ISSUE THREAD

**Issue by kate-goldenring** [bytecodealliance/component-docs#182]
The book is lacking in an example and more documentation on using WASI HTTP to create Wasm services and running them with `cargo serve` and `wasmtime serve` etc. Instead of writing and maintaining an example, @yoshuawuyts has [started the process](https://github.com/bytecodealliance/governance/issues/113) of upstreaming his [wasi:http Rust sample](https://github.com/yoshuawuyts/sample-wasi-http-rust) to the BA. We should create a section with this new example. Ideally, we can show composition with this component too with registries

**Comment by kate-goldenring** [comment]
To close this: 
1. Find a place to put this example - potentially add a new examples section or a WASI HTTP tutorial
2. Include a way to get started with the example

Recommended approach: examples section under "Using the Component Model"

**Comment by hakilebara** [comment]
I drafted a WASI HTTP tutorial ([here](https://www.notion.so/Writing-an-HTTP-Wasm-Component-in-Rust-33ea97bc458a804094c6c9beb78b9a4e)). Could it fit the requirement of this issue?

**Comment by vados-cosmonic** [comment]
Hey @hakilebara it could definitely fit the requirement of the issue -- I think that ideally we'd want something like a new section (section 6.5) with how to use HTTP in components. Your guide could be the first subsection under that, dedicated to how to do it for Rust! 

If you'd like to please feel free to make a PR and we'll get around to reviewing it.

**Comment by hakilebara** [comment]
> I think that ideally we'd want something like a new section (section 6.5) with how to use HTTP in components

Noted thank you. So something like this, where the page `Using HTTP in Components` contains a summary of the section and a list of its sub-sections?

```diff
 ├── Creating Components
 │   └── Building a simple component
 │   └── Importing and reusing components
 │   └── Creating Runnable Components
 │   └── Using WIT resources
+│   ├── Using HTTP in Components
+|   |   └── Rust
```

Should the guide explicitly generate bindings or should it use [wstd](https://github.com/bytecodealliance/wstd)?

I will try and make a PR.

**Comment by vados-cosmonic** [comment]
Yeah, that's exactly the right idea -- for now I think that's a the right (and other maintainers of component-docs should be able to chime in on the PR!)

> Should the guide explicitly generate bindings or should it use [wstd](https://github.com/bytecodealliance/wstd)?

This is a very good question -- I'd like to think that we should show `wstd` first, and then have a section (you do not have to write this section, unless you'd really like to!) that does it "the hard way".

**Comment by kate-goldenring** [comment]
Thanks @hakilebara. I like the tutorial. My main concern is maintainability. I was preferential to mainly referencing the BA example repos because they are well maintained: https://github.com/bytecodealliance/sample-wasi-http-rust. Maybe we can link to them at the top of the tutorial and make sure yours has a similar code implementation

**Comment by hakilebara** [comment]
Thanks for the feedback @kate-goldenring. My mental model when writing the doc was to give the minimum amount of code to curl a "hello world" from localhost. The doc would then end with a `Going further` section that includes [bytecodealliance/sample-wasi-http-rust](https://github.com/bytecodealliance/sample-wasi-http-rust) in a list of repos that implement HTTP components in Rust.

<br />

> Maybe we can link to them at the top of the tutorial and make sure yours has a similar code implementation
 
The doc (`Using HTTP in Components/Rust.md`) and the repo [bytecodealliance/sample-wasi-http-rust](https://github.com/bytecodealliance/sample-wasi-http-rust) have in common the use of `wstd` thus the code of the handlers is quite similar. However, they do have the following differences:
- The repo was created as a Rust library project, while the doc isn't. The component code is therefore stored in `src/lib.rs` and `src/main.rs` respectively. This is explained in the doc but could throw some readers off
- The repo creates a WIT file and fetches dependencies while the doc don't

<br />

I understand and share you concern on maintainability. What I can propose is to either :
- adapt the doc to closely match the repo
- drop the doc and transform the section page `Using HTTP in Components` into a reference list of the BA example repos

**Comment by kate-goldenring** [comment]
@hakilebara the doc you created is very streamlined such that i was able to walk through it in a minute. that assuages some of my concerns about maintainability. I think it is fine how you've done it, and i appreciate the going further section. If you're willing, I'd love to have you contribute again for as the doc needs updates.
